### PR TITLE
chore: additional metadata for protocol capture errors

### PR DESCRIPTION
**Pull request by cacieprins**
This attaches additional metadata to non-fatal protocol capture errors, and reports all fatal protocol capture errors.

### Additional details

In order to better collect error metadata in sentry for protocol capture workflows, additional context is reported for non-fatal errors that occur during capture. Fatal errors, such as errors initializing the protocol capture code, are reported via the artifacts endpoint for more decisive reporting. Error handling for downloading and initializing the capture code was improved in order to accomplish this. 

### PR Tasks

- [X] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/server/lib/cloud/api.ts [hunk 1]
```diff
@@ -384,38 +384,35 @@ module.exports = {
...
-            },
-          ])
+          protocolManager.addFatalError('getCaptureProtocolScript', error, [result.captureProtocolUrl])
         } else {
           throw e
         }
       }
 
+      if (script) {
+        await options.project.protocolManager.setupProtocol(script, result.runId)
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/api.ts]: What if `setupProtocol` throws? We should treat that as a fatal error as well right?

Comment (cacieprins) [reply packages/server/lib/cloud/api.ts]: Yep, that try/catch is already within protocolManager. I'll add a test for it

File: packages/server/lib/cloud/protocol.ts [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
 import Database from 'better-sqlite3'
 import path from 'path'
 import os from 'os'
 import { createGzip } from 'zlib'
 import fetch from 'cross-fetch'
 import Module from 'module'
 import env from '../util/env'
-
+import base64url from 'base64url'
+import crypto from 'crypto'
```
Comment (AtofStryker) [inline packages/server/lib/cloud/protocol.ts]: nit: can we organize the imports to be from node/modules, then src packages, then types from src packages?

File: packages/server/lib/cloud/protocol.ts [hunk 3]
```diff
@@ -344,7 +395,7 @@ export class ProtocolManager implements ProtocolManagerShape {
    * Abstracts invoking a synchronous method on the AppCaptureProtocol instance, so we can handle
    * errors in a uniform way
    */
-  private async invokeAsync <K extends ProtocolAsyncMethods> (method: K, ...args: Parameters<AppCaptureProtocolInterface[K]>) {
+  private async invokeAsync <K extends ProtocolAsyncMethods> (method: K, isEssential: boolean, ...args: Parameters<AppCaptureProtocolInterface[K]>) {
```
Comment (AtofStryker) [inline packages/server/lib/cloud/protocol.ts]: can we make the second option an object to make the intention of the boolean a bit clearer without necessarily having to look up the method? like:

```
this.invokeAsync('beforeTest', { isEssential: true }, test)
```

Comment (cacieprins) [reply packages/server/lib/cloud/protocol.ts]: Considered it, and thought vscode's hinting would be enough. I'll change it up.

File: packages/server/lib/cloud/api.ts (hunk 1 above)
Comment (cacieprins) [reply packages/server/lib/cloud/api.ts]: Actually - setupProtocol is only in the app side of protocol manager; the constructor error state tests this case

Comment (ryanthemanuel) [reply packages/server/lib/cloud/api.ts]: I see now 👍🏻

**Review by ryanthemanuel — APPROVED** [review]

**Review by AtofStryker — APPROVED** [review]

**Pull request merged**